extensions/HyperVExtension/src/HyperVExtension/Models/VirtualMachineCreation/FileDownloadMonitor.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using HyperVExtension.Extensions;
+using Serilog;
+
+namespace HyperVExtension.Models.VirtualMachineCreation;
+
+/// <summary>
+/// Monitors the download of a file and allows clients to subscribe to the progress of the
+/// download. As new progress comes in the FileDownloadMonitor publishes this progress to
+/// its subscribers.
+/// </summary>
+internal sealed class FileDownloadMonitor
+{
+    private readonly ILogger _log = Log.ForContext("SourceContext", nameof(FileDownloadMonitor));
+
+    private readonly object _lock = new();
+
+    private readonly List<IProgress<IOperationReport>> _subscriberList = new();
+
+    private readonly Progress<ByteTransferProgress> _progressReporter;
+
+    private bool _downloadInProgress;
+
+    private DownloadOperationReport? _lastSentReport;
+
+    public FileDownloadMonitor(IProgress<IOperationReport> progressProvider)
+    {
+        AddSubscriber(progressProvider);
+        _progressReporter = new(PublishProgress);
+    }
+
+    public void AddSubscriber(IProgress<IOperationReport> progressProvider)
+    {
+        lock (_lock)
+        {
+            if (_lastSentReport != null)
+            {
+                // Subscriber addition requested so we'll submit the last recorded
+                // progress we received before adding it to the subscriber list
+                progressProvider.Report(_lastSentReport);
+            }
+
+            _subscriberList.Add(progressProvider);
+        }
+    }
+
+    private void PublishProgress(ByteTransferProgress transferProgress)
+    {
+        _lastSentReport = new DownloadOperationReport(transferProgress);
+        _subscriberList.ForEach(subscriber => subscriber.Report(_lastSentReport));

```

Comment:
We are traversing _subscriberList here without locking while it can be modified from AddSubscriber.

Reply:
ah, missed it since I call it in the StartAsync method below. But makes sense since this is getting called everytime there is a progress update outside of StartAsync. Will update this.